// @angular/cli/blueprints/directive/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ast_tools_1 = require("../../lib/ast-tools");
var config_1 = require("../../models/config");
var app_utils_1 = require("../../utilities/app-utils");
var path = require('path');
var fs = require('fs');
var chalk = require('chalk');
var dynamicPathParser = require('../../utilities/dynamic-path-parser');
var stringUtils = require('ember-cli-string-utils');
var astUtils = require('../../utilities/ast-utils');
var findParentModule = require('../../utilities/find-parent-module').default;
var Blueprint = require('../../ember-cli/lib/models/blueprint');
var getFiles = Blueprint.prototype.files;
exports.default = Blueprint.extend({
    description: '',
    aliases: ['d'],
    availableOptions: [
        {
            name: 'flat',
            type: Boolean,
            description: 'Flag to indicate if a dir is created.'
        },
        {
            name: 'prefix',
            type: String,
            default: null,
            description: 'Specifies whether to use the prefix.'
        },
        {
            name: 'spec',
            type: Boolean,
            description: 'Specifies if a spec file is generated.'
        },
        {
            name: 'skip-import',
            type: Boolean,
            default: false,
            description: 'Allows for skipping the module import.'
        },
        {
            name: 'module',
            type: String, aliases: ['m'],
            description: 'Allows specification of the declaring module.'
        },
        {
            name: 'export',
            type: Boolean,
            default: false,
            description: 'Specifies if declaring module exports the component.'
        },
        {
            name: 'app',
            type: String,
            aliases: ['a'],
            description: 'Specifies app name to use.'
        }
    ],
    beforeInstall: function (options) {
        var appConfig = app_utils_1.getAppFromConfig(this.options.app);
        if (options.module) {
            // Resolve path to module
            var modulePath = options.module.endsWith('.ts') ? options.module : options.module + ".ts";
            var parsedPath = dynamicPathParser(this.project, modulePath, appConfig);
            this.pathToModule = path.join(this.project.root, parsedPath.dir, parsedPath.base);
            if (!fs.existsSync(this.pathToModule)) {
                throw ' ';
            }
        }
        else {
            try {
                this.pathToModule = findParentModule(this.project.root, appConfig.root, this.generatePath);
            }
            catch (e) {
                if (!options.skipImport) {
                    throw "Error locating module for declaration\n\t" + e;
                }
            }
        }
    },
    normalizeEntityName: function (entityName) {
        var appConfig = app_utils_1.getAppFromConfig(this.options.app);
        var parsedPath = dynamicPathParser(this.project, entityName, appConfig);
        this.dynamicPath = parsedPath;
        var defaultPrefix = (appConfig && appConfig.prefix) || '';
        var prefix = (this.options.prefix === 'false' || this.options.prefix === '')
            ? '' : (this.options.prefix || defaultPrefix);
        prefix = prefix && prefix + "-";
        this.selector = stringUtils.camelize(prefix + parsedPath.name);
        return parsedPath.name;
    },
    locals: function (options) {
        options.spec = options.spec !== undefined ?
            options.spec : config_1.CliConfig.getValue('defaults.directive.spec');
        options.flat = options.flat !== undefined ?
            options.flat : config_1.CliConfig.getValue('defaults.directive.flat');
        return {
            dynamicPath: this.dynamicPath.dir,
            flat: options.flat,
            selector: this.selector
        };
    },
    files: function () {
        var fileList = getFiles.call(this);
        if (this.options && !this.options.spec) {
            fileList = fileList.filter(function (p) { return p.indexOf('__name__.directive.spec.ts') < 0; });
        }
        return fileList;
    },
    fileMapTokens: function (options) {
        var _this = this;
        // Return custom template variables here.
        return {
            __path__: function () {
                var dir = _this.dynamicPath.dir;
                if (!options.locals.flat) {
                    dir += path.sep + options.dasherizedModuleName;
                }
                _this.generatePath = dir;
                return dir;
            }
        };
    },
    afterInstall: function (options) {
        var _this = this;
        var returns = [];
        var className = stringUtils.classify(options.entity.name + "Directive");
        var fileName = stringUtils.dasherize(options.entity.name + ".directive");
        var fullGeneratePath = path.join(this.project.root, this.generatePath);
        var moduleDir = path.parse(this.pathToModule).dir;
        var relativeDir = path.relative(moduleDir, fullGeneratePath);
        var importPath = relativeDir ? "./" + relativeDir + "/" + fileName : "./" + fileName;
        if (!options.skipImport) {
            if (options.dryRun) {
                this._writeStatusToUI(chalk.yellow, 'update', path.relative(this.project.root, this.pathToModule));
                return;
            }
            returns.push(astUtils.addDeclarationToModule(this.pathToModule, className, importPath)
                .then(function (change) { return change.apply(ast_tools_1.NodeHost); })
                .then(function (result) {
                if (options.export) {
                    return astUtils.addExportToModule(_this.pathToModule, className, importPath)
                        .then(function (change) { return change.apply(ast_tools_1.NodeHost); });
                }
                return result;
            }));
            this._writeStatusToUI(chalk.yellow, 'update', path.relative(this.project.root, this.pathToModule));
        }
        return Promise.all(returns);
    }
});
